crates/daphne-worker/src/aggregator/queues/async_aggregator.rs

```diff
@@ -0,0 +1,342 @@
+                .request(Command::Check, (task_id, epoch, shard))
+                .encode(&replay_checker::Request {
+                    report_ids: report_ids.into(),
+                    aggregation_job_id: agg_job_id,
+                })
+                .send::<replay_checker::Response>()
+                .await
+                .map(|r| r.duplicates)
+        })
+        .buffer_unordered(6)
+        .try_fold(HashSet::new(), |mut acc, dups| async move {
+            acc.extend(dups);
+            Ok(acc)
+        })
+        .await
+}
+
+macro_rules! bail {
+    (retry $m:ident, err = $error:expr, $msg:literal) => {{
+        $m.retry();
+        bail!(err = $error, $msg)
+    }};
+    (err = $error:expr, $msg:literal) => {{
+        tracing::error!(error = ?$error, $msg);
+        return;
+    }}
+}
+
+/// Perform an aggregation job.
+///
+/// ## Note
+/// There is a worst case scenario that this handler can't deal with.
+///
+/// Messages will be replayed if they fail, but the workers runtime will eventually give up on a
+/// message after it's been retried a bunch of times, in that case the helper will never respond
+/// positively to the poll request, possibly leaving the leader in an infinite loop state. The
+/// leader can resubmit the work as many times as it wants to get out of this situation, but
+/// implementers of the leader must be made aware of this.
+//
+// -----
+//
+// All of the IO in this function is idempotent. They can be spotted by looking at the `.await`
+// expressions. The explanation is as follows:
+//  1. Getting a task config. No writes performed.
+//  2. Initializing the reports. Stateless.
+//  3. The same aggregation job may replay it's own reports, this means replay checking is
+//     idempotent. See the [ReplayChecker] durable object for more details.
+//  4. Storing the aggregate share does not merge with any other aggregate share and simply
+//     replaces the previous one, which will be identical.
+//  5. Storing the aggregate response simply overwrites the previous response, which will be
+//     identical.
+//
+#[tracing::instrument(skip_all, fields(?version, ?part_batch_sel, ?agg_job_id))]
+pub async fn async_aggregate_one(
+    app: &App,
+    message: &RawMessage,
+    AsyncAggregationMessage {
+        version,
+        part_batch_sel,
+        agg_job_id,
+        initialize_reports,
+        taskprov_advertisement,
+    }: AsyncAggregationMessage<'_>,
+) {
+    let task_id = &initialize_reports.task_id;
+    tracing::debug!("getting task config");
+    let taskprov_advertisement = match taskprov_advertisement
+        .map(|s| TaskprovAdvertisement::parse_taskprov_advertisement(&s, task_id, version))
+        .transpose()
+    {
+        Ok(taskprov_advertisement) => taskprov_advertisement,
+        Err(e) => bail!(err = e, "taskprov advertisement was illformed"),
+    };
+
+    // 1.
+    let (task_config, get_task_config_duration) = time_it(daphne::roles::resolve_task_config(
```

Comment:
Maybe a little clearer
```suggestion
    let (task_config, get_task_config_duration) = timed(daphne::roles::resolve_task_config(
```